### ISSUE TITLE
The implementation is really unsound, because of incorrect assumptions.

### ISSUE THREAD

**Issue by TernaryM01** [jonhoo/roget#29]
First of all, I would like to thank Jon very much for his fantastic YouTube channel which allowed me to very quickly learn programming in Rust. This project, in particular, is the perfect kind of project for me in learning a programming language: Tackle a practical challenging problem, and optimize the code as much as possible.

Unfortunately, Jon became focused on optimizing his code without verifying that his logic is sound. The program becomes faster and faster and more and more wrong. First, he didn't realize that he forced his solver to only work in hard mode (every guess has to be a candidate of the right answer based on the correctness patterns received). Second, memoization can only be done for the specific history of guesses and the corresponding received correctness patterns since the first guess, but he thought that he could localize it. And from there it spiraled all the way down that it became impossible for me to track his reasoning throughout his live-coding video.

So, I decided to re-do roget from the bottom up, following only the mathematically sound ideas and the good programming patterns. It was first intended to be just a private project to learn Rust, but I decided to publish it on GitHub. You can find it [here](https://github.com/TernaryM01/rogerthat). It can be regarded as a fork of an early version of roget, but I typed it from scratch while following Jon's recorded video, making my own adjustments here and there.

Ironically for Jon, he missed **_an opportunity to gain a quadratic speedup_** by decoupling the for loop iterating through guess candidates and the for loop iterating through correctness patterns, observing that every guess-answer pair corresponds to exactly one correctness pattern. This dramatic improvement is wholly sound and doesn't change the guessing behavior at all. You can check out the implementation of this in [rogerthat](https://github.com/TernaryM01/rogerthat).

**Comment by jonhoo** [comment]
Hi there! Glad you enjoyed the beginning of the video at least :)

To be clear, many of the issues you point out have already been pointed out and fixes since in various issues and PRs on this repo, and the solver is now _way_ faster. Definitely true that this is an indicator of just how careful one has to be while pursuing optimizations!

**Comment by TernaryM01** [comment]
Thanks for your quick response! No worries, I still watched the video until the end to see the workflow of a professional Rust programmer, including how you use profiling tools to analyze performance (and what lessons I could take from the shortcomings as well 😋). And I enjoy all of your Crust of Rust series, getting into the nitty-gritty of how Rust works! Those are YouTube gems! 😄

Despite of what you said, however, I would like to believe that I have done my due diligence to read the repo as it stands today and to track all the development throughout all the commits and the PRs. It seems to me that many of the problematic aspects still haven't been mitigated (except [normal mode](https://github.com/jonhoo/roget/commit/4482ad13db3c556703565f81419223f365c4f96b), being introduced so late when the state of the algorithm had already become so unsound), ~~and the quadratic speedup I've mentioned hasn't been implemented.~~ (Correction: Turns out it has been introduced by cjcormier [since this PR.](https://github.com/jonhoo/roget/pull/14) My bad...) Otherwise, I would've decided to just build on top of what's currently there and contributed to this repo instead of creating my own.

Roget is currently _way_ faster indeed than the current state of rogerthat. The problem is on whether it's _sound_. I've tried and will continue to pull ideas from roget to make every bit of performance improvement possible on my implementation. It just so happens that so many of them turn out to be unsound, so I can't use them. (Much of the contribution from other people just built on top of what's already there, assuming that everything's good so far.)

But perhaps it would be better for everyone if I discuss each of those particular problems in separate issues/comments, so that I can go into detail on the specific portions.

As a postscript, to be careful in making improvements to a Wordle solver, one has to carefully track its change of behavior (i.e. what guesses it makes, not how quickly it makes them). Sophisticated test cases and hands-on analysis is required. Overall statistics can be deceiving, in that adding a bunch of clever heuristics and even cheating by overfitting parameters on the answers set will typically only reduce the average number of guesses like from 3.7 to 3.6 (that's a lesson I learned from 3blue1brown).
